sys/kern/syscalls.c

```diff
@@ -896,4 +896,50 @@ static int sys_fstatvfs(proc_t *p, fstatvfs_args_t *args, register_t *res) {
     error = copyout_s(buf, u_buf);
 
   return error;
-}
\ No newline at end of file
+}
+
+static int sys_getgroups(proc_t *p, getgroups_args_t *args, register_t *res) {
+  int error = 0;
+  int ngroups = args->ngroups;
+  gid_t *ugidset = args->gidset;
+  int pngroups = p->p_cred.cr_ngroups;
+
+  klog("getgroups(%d, %p)", ngroups, ugidset);
+
+  if (ngroups == 0) {
+    /* just return number of groups */
+    *res = pngroups;
+    return 0;
+  }
+
+  if (ngroups < pngroups)
+    return EINVAL;
+
+  *res = pngroups;
+  error = copyout(p->p_cred.cr_groups, ugidset, pngroups * sizeof(gid_t));
+  return error;
+}
+
+static int sys_setgroups(proc_t *p, setgroups_args_t *args, register_t *res) {
+  int error = 0;
+  int ungroups = args->ngroups;
+  const gid_t *ugidset = args->gidset;
+
+  klog("setgroups(%d, %p)", ungroups, ugidset);
+
+  /* too many groups */
+  if (ungroups > NGROUPS_MAX)
```

Comment:
A malicious user may set `ngroups` to negative number. Have you handled that correctly?

Reply:
Changed. I forgot about it.